src/state.rs

```diff
@@ -1,46 +1,120 @@
+use tui::widgets::ListState;
 
 use crate::tasks::Task;
 
 use std::fs::{self, File, OpenOptions};
 
-#[derive(Default)]
 pub struct State {
-    tasks: Vec<Task>,
+    pub text_input: String,
+    pub tasks: StatefulList<Task>,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        State {
+            text_input: String::new(),
+            tasks: StatefulList::with_items(Vec::<Task>::new()),
+        }
+    }
 }
 
 impl State {
-    pub fn load_tasks(&mut self) -> std::io::Result<()> {
-        let file = open_file()?;
+    pub fn load_tasks(&mut self) -> Result<()> {
+        let file = open_file(FileMode::Load)?;
         let tasks: Vec<Task> = serde_json::from_reader(file)?;
-        self.tasks = tasks;
+        self.tasks = StatefulList::with_items(tasks);
         Ok(())
     }
 
-    pub fn save_tasks(&self) -> std::io::Result<()> {
-        let file = open_file()?;
-        serde_json::to_writer_pretty(file, &self.tasks)?;
+    pub fn save_tasks(&self) -> Result<()> {
+        let file = open_file(FileMode::Save)?;
+        serde_json::to_writer_pretty(file, &self.tasks.items)?;
         Ok(())
     }
 
-    pub fn add_task(&mut self, name: &str, priority: Option<u8>, date: Option<DateTime<Local>>) {
-        let task = Task::new(name, priority, date);
-        self.tasks.push(task);
+    pub fn add_task(&mut self) {
+        let task = Task::from_string(&self.text_input).unwrap(); // TODO: handle failure
+        self.tasks.items.push(task);
+        self.save_tasks()
+            .expect("error: could not save tasks to a file"); // TODO: handle failure
+        self.text_input = String::new();
     }
 }
 
-fn open_file() -> std::io::Result<File> {
-    let file = OpenOptions::new()
-        .read(true)
-        .write(true)
-        .create(true)
-        .open("tasks.json")?;
-    if fs::metadata("tasks.json")?.len() == 0 {
-        let default_content: Vec<Task> = Vec::new();
-        fs::write(
-            "tasks.json",
-            serde_json::to_string_pretty(&default_content)?,
-        )?
+enum FileMode {
+    Load,
+    Save,
+}
+
+fn open_file(mode: FileMode) -> Result<File> {
+    let file = match mode {
+        FileMode::Load => {
+            let f = OpenOptions::new()
+                .write(true)
+                .create(true)
+                .read(true)
+                .open("tasks.json")?;
+            if fs::metadata("tasks.json")?.len() == 0 {
+                let default_content: Vec<Task> = Vec::new();
+                fs::write(
+                    "tasks.json",
+                    serde_json::to_string_pretty(&default_content)?,
+                )?
+            };
+            f
+        }
+        FileMode::Save => OpenOptions::new()
+            .write(true)
+            .truncate(true)
+            .create(true)
+            .open("tasks.json")?,
     };
     Ok(file)
 }
+
+pub struct StatefulList<T> {
+    pub state: ListState,
+    pub items: Vec<T>,
+}
+
+impl<T> StatefulList<T> {
```

Comment:
Do we need this wrapper? What about storing index of currently selected task, seems a lot easier and cleaner imo. If you dont plan to expand functionality of this class later then i suggest the proposed solution :)

Reply:
We would need ListState associated with list items so we can render_stateful_widget(), but I'a open to this proposition.